src/backend/distributed/transaction/backend_data.c

```diff
@@ -0,0 +1,480 @@
+
+
+/*
+ * assign_distributed_transaction_id updates the shared memory allocated for this backend
+ * and sets initiatorNodeIdentifier, transactionId, timestamp fields with the given
+ * inputs. Also, the function sets the database id and process id via the information that
+ * Postgres provides.
+ *
+ * This function is only intended for internal use for managing distributed transactions.
+ * Users should not use this function for any purpose.
+ */
+Datum
+assign_distributed_transaction_id(PG_FUNCTION_ARGS)
+{
+	CheckCitusVersion(ERROR);
+
+	/* MyBackendData should always be avaliable, just out of paranoia */
+	if (!MyBackendData)
+	{
+		ereport(ERROR, (errmsg("backend is not ready for distributed transactions")));
+	}
+
+	SpinLockAcquire(&MyBackendData->mutex);
+
+	/* if an id is already assigned, relase the lock and error */
+	if (MyBackendData->transactionId.initiatorNodeIdentifier != 0)
+	{
+		SpinLockRelease(&MyBackendData->mutex);
+
+		ereport(ERROR, (errmsg("the backend has already been assigned a "
+							   "transaction id")));
+	}
+
+	MyBackendData->databaseId = MyDatabaseId;
+
+	MyBackendData->transactionId.initiatorNodeIdentifier = PG_GETARG_INT32(0);
+	MyBackendData->transactionId.transactionNumber = PG_GETARG_INT64(1);
+	MyBackendData->transactionId.timestamp = PG_GETARG_TIMESTAMPTZ(2);
+
+	SpinLockRelease(&MyBackendData->mutex);
+
+	PG_RETURN_VOID();
+}
+
+
+/*
+ * get_current_transaction_id returns a tuple with (databaseId, processId,
+ * initiatorNodeIdentifier, transactionNumber, timestamp) that exists in the
+ * shared memory associated with this backend. Note that if the backend
+ * is not in a transaction, the function returns uninitialized data where
+ * transactionNumber equals to 0.
+ */
+Datum
+get_current_transaction_id(PG_FUNCTION_ARGS)
+{
+	TupleDesc tupleDescriptor = NULL;
+	HeapTuple heapTuple = NULL;
+
+	const int attributeCount = 5;
+	Datum values[attributeCount];
+	bool isNulls[attributeCount];
+
+	DistributedTransactionId *distributedTransctionId = NULL;
+
+	CheckCitusVersion(ERROR);
+
+	/* build a tuple descriptor for our result type */
+	if (get_call_result_type(fcinfo, NULL, &tupleDescriptor) != TYPEFUNC_COMPOSITE)
+	{
+		elog(ERROR, "return type must be a row type");
+	}
+
+	/* MyBackendData should always be avaliable, just out of paranoia */
+	if (!MyBackendData)
+	{
+		ereport(ERROR, (errmsg("backend is not ready for distributed transactions")));
+	}
+
+	distributedTransctionId = GetCurrentDistributedTransctionId();
+
+	memset(values, 0, sizeof(values));
+	memset(isNulls, false, sizeof(isNulls));
+
+	/* first two fields do not change for this backend, so get directly */
+	values[0] = ObjectIdGetDatum(MyDatabaseId);
+	values[1] = Int32GetDatum(MyProcPid);
+
+	values[2] = Int32GetDatum(distributedTransctionId->initiatorNodeIdentifier);
+	values[3] = UInt64GetDatum(distributedTransctionId->transactionNumber);
+	values[4] = TimestampTzGetDatum(distributedTransctionId->timestamp);
+
```

Comment:
I think we should return NULL when there's no distributed transaction in progress. Showing a random date in 1999 is a bit confusing / not helpful.